libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsPrefetcher.kt

```diff
@@ -0,0 +1,168 @@
+
+    private val routeProgressObserver = RouteProgressObserver { onRouteProgressChanged(it) }
+
+    private val ignoredRouteUpdateReasons = setOf(
+        RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP,
+        RoutesExtra.ROUTES_UPDATE_REASON_ALTERNATIVE,
+        RoutesExtra.ROUTES_UPDATE_REASON_REFRESH,
+    )
+    private var lastDownloadTime: Long = 0
+
+    /**
+     * See [MapboxNavigationObserver.onAttached].
+     */
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.registerRoutesObserver(routesObserver)
+        mapboxNavigation.registerRouteProgressObserver(routeProgressObserver)
+    }
+
+    /**
+     * See [MapboxNavigationObserver.onDetached].
+     */
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.unregisterRoutesObserver(routesObserver)
+        mapboxNavigation.unregisterRouteProgressObserver(routeProgressObserver)
+    }
+
+    private fun onRoutesChanged(result: RoutesUpdatedResult) {
+        if (result.reason in ignoredRouteUpdateReasons) {
+            return
+        }
+        result.navigationRoutes.firstOrNull()?.let {
+            val currentStep = it.directionsRoute.legs()?.firstOrNull()?.steps()?.firstOrNull()
+            if (currentStep != null) {
+                val progress = RouteProgressData(
+                    it.directionsRoute,
+                    0,
+                    0,
+                    currentStep.duration(),
+                    currentStep.distance()
+                )
+                triggerDownload(progress)
+            }
+        }
+    }
+
+    private fun onRouteProgressChanged(routeProgress: RouteProgress) {
+        if (shouldDownloadBasedOnTime()) {
+            val legIndex = routeProgress.currentLegProgress?.legIndex
+            val currentStepProgress = routeProgress.currentLegProgress?.currentStepProgress
+            val stepIndex = currentStepProgress?.stepIndex
+            val durationRemaining = currentStepProgress?.durationRemaining
+            val distanceRemaining = currentStepProgress?.distanceRemaining
+            ifNonNull(
+                legIndex,
+                stepIndex,
+                durationRemaining,
+                distanceRemaining
+            ) { legIndex, stepIndex, durationRemaining, distanceRemaining ->
+                val progressData = RouteProgressData(
+                    routeProgress.route,
+                    legIndex,
+                    stepIndex,
+                    durationRemaining,
+                    distanceRemaining.toDouble()
+                )
+                triggerDownload(progressData)
+            }
+        }
+    }
+
+    /**
+     * The method stops all work related to pre-downloading voice instructions and unregisters
+     * all related callbacks. It should be invoked from `Activity#onDestroy`.
+     */
+    fun destroy() {
+        speechApi.destroy()
+    }
```

Comment:
Activity is not always the owner for these types of objects, so I think we should delete this function in favor of cleaning up resources inside the `onDetached`

Reply:
I think you're right. But does this mean that `onAttached`/`onDetached` should be invoked from `onCreate`/`onDestroy`? I think yes. Could you please review `MapboxVoiceActivity` and confirm?